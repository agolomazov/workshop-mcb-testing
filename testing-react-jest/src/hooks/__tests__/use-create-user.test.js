import { renderHook, act } from '@testing-library/react'
import { useCreateUser } from '../use-create-user'
import { passwordValidationErrors } from '../../constants/validation';

describe('useCreateUser()', () => {
  it('Вернется объект с корректными полями', () => {
    const { result } = renderHook(useCreateUser);

    expect(result.current).toHaveProperty('successMessage');
    expect(result.current).toHaveProperty('errorMessage');
    expect(result.current).toHaveProperty('onSuccess');
    expect(result.current).toHaveProperty('onError');
    expect(result.current).toHaveProperty('onSubmit');
  });

  it('Проверяем корректность типов полей хука', () => {
    const { result } = renderHook(useCreateUser);

    expect(typeof result.current.successMessage).toBe('string');
    expect(typeof result.current.errorMessage).toBe('string');
    expect(typeof result.current.onSubmit).toBe('function');
    expect(typeof result.current.onError).toBe('function');
    expect(typeof result.current.onSuccess).toBe('function');
  });

  it('Проверяем установку successMessage', () => {
    const { result } = renderHook(useCreateUser);

    expect(result.current.successMessage).toBe('');

    act(() => {
      result.current.onSuccess({ name: 'Anton', password: 'qweTR123!!##' });
    });

    expect(result.current.successMessage).toMatchSnapshot();
  });

  it('Проверяем установку errorMessage', () => {
    const errorMessage = 'shit happens';
    const { result } = renderHook(useCreateUser);

    expect(result.current.errorMessage).toBe('');

    act(() => {
      result.current.onError(new Error(errorMessage));
    });

    expect(result.current.errorMessage).toBe(errorMessage);
  });

  it('Проверка выброса ошибки - короткий пароль', async () => {
    const { result } = renderHook(useCreateUser);

    await expect(result.current.onSubmit({ password: '123' })).rejects.toThrow(passwordValidationErrors.length);
  })

  it('Проверка выброса ошибки - неправильный регистр символов', async () => {
    const { result } = renderHook(useCreateUser);

    await expect(result.current.onSubmit({ password: '1234567890' })).rejects.toThrow(
      passwordValidationErrors.case,
    );
  });

  it('Проверка выброса ошибки - нет цифр', async () => {
    const { result } = renderHook(useCreateUser);

    await expect(
      result.current.onSubmit({ password: 'asfjADASDlkjd##!@' }),
    ).rejects.toThrow(passwordValidationErrors.number);
  });

  it('Проверка выброса ошибки - нет спецсимволов', async () => {
    const { result } = renderHook(useCreateUser);

    await expect(
      result.current.onSubmit({ password: 'asfjADASDlkj90909' }),
    ).rejects.toThrow(passwordValidationErrors.special);
  });

  it('Проверка выброса ошибки - ошибок нет', async () => {
    const { result } = renderHook(useCreateUser);

    await expect(
      result.current.onSubmit({ password: 'asfjADASDlkj9090№№9' }),
    ).resolves.toBe();
  });
})