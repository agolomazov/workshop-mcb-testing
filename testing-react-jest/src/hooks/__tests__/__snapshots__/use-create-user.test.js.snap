// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCreateUser() Проверяем установку successMessage 1`] = `"User Anton created with password qweTR123!!##"`;
